### PR TITLE
484/zap flow

### PR DESCRIPTION
**Pull request by Rudi114**
#484

### CODE REVIEW

**Comment by alex0xbe0** [comment]
Merge conflicts

File: src/components/custom/create-position-bond/index.tsx [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+          <Balance
+            title={`Deposit ${stateMachine.context.tokenSymbol}`}
+            value={`Balance: ${tokenInfo?.humanValue?.toFixed()}`}
+          />
+          <Form.Item name="tokenAmount" required>
+            <TokenAmount
+              disabled={loading}
+              displayDecimals={tokenInfo?.decimals}
+              mainAsset={collateral.vault.name}
+              max={tokenInfo?.humanValue}
```
Comment (whoislewys) [inline src/components/custom/create-position-bond/index.tsx:192]: i'm unable to mint fiat from either create position tab (bond or underlying) - a max of 0 is displayed. maybe an issue snuck in with this `max` value when you were moving components around?

File: src/constants/misc.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ export const SET_ALLOWANCE_PROXY_TEXT = 'Set allowance for Proxy'
 export const ENABLE_PROXY_FOR_FIAT_TEXT = 'Enable Proxy for FIAT'
 export const EXECUTE_TEXT = 'Execute'
 export const DEPOSIT_COLLATERAL_TEXT = 'Deposit collateral'
+export const DEPOSIT_UNDERLYING_TEXT = 'Deposit Underlying'
```
Comment (whoislewys) [inline src/constants/misc.ts:40]:
```suggestion
export const DEPOSIT_UNDERLYING_TEXT = 'Deposit underlying'
```

File: pages/create-position/[collateralId]/open/index.tsx [hunk 3]
```diff
@@ -214,155 +169,54 @@ const FormERC20: React.FC<{
...
+                  Underlying
+                </RadioTab>
+              </RadioTabsWrapper>
+            )}
+            {tab === 'underlying' ? (
+              <CreatePositionUnderlying
+                collateral={collateral}
+                loading={loading}
+                healthFactorNumber={healthFactorNumber}
+                maxBorrowAmountCalculated={maxBorrowAmountCalculated}
```
Comment (whoislewys) [inline pages/create-position/[collateralId]/open/index.tsx:204]: unable to take out fiat loans, there is a max borrow of 0. maybe this `maxBorrowAmount` is calculating the wrong thing?

File: src/components/custom/swap-settings-modal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+interface Props {
+  isOpen: boolean
+  toggleOpen: () => void;
+  submit: (slippageTolerance: number, maxTransactionTime: number) => void
+}
+
+const SwapSettingsModal: React.FC<Props> = ({isOpen, toggleOpen, submit}: Props) => {
+
+  const [slippageTolerance, setSlippageTolerance] = useState(0.1)
```
[the remark this answers is not in the document]
Comment (Rudi114) [reply src/components/custom/swap-settings-modal/index.tsx]: 🤘

File: src/components/custom/swap-settings-modal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  const [maxTransactionTime, setMaxTransactionTime] = useState(20)
+
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (Rudi114) [reply src/components/custom/swap-settings-modal/index.tsx]: 👍

File: src/components/custom/swap-settings-modal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import Modal                           from '@/src/components/antd/modal'
+import { Text }                        from '@/src/components/custom/typography'
+import ButtonGradient                  from '@/src/components/antd/button-gradient'
+import ButtonOutlineGradient           from '@/src/components/antd/button-outline-gradient'
+import { InputNumber }                 from 'antd';
+
+interface Props {
+  isOpen: boolean
+  toggleOpen: () => void;
+  submit: (slippageTolerance: number, maxTransactionTime: number) => void
```
[the remark this answers is not in the document]
Comment (Rudi114) [reply src/components/custom/swap-settings-modal/index.tsx]: 👍

File: src/hooks/useTokenDecimalsAndBalance.tsx [hunk 7]
```diff
@@ -23,7 +23,7 @@ export const useTokenDecimalsAndBalance = ({
   vaultType,
 }: {
   tokenId: string
-  vaultType: string
+  vaultType?: string
```
[the remark this answers is not in the document]
Comment (Rudi114) [reply src/hooks/useTokenDecimalsAndBalance.tsx:26]: vaultType is used to determine whether to return an erc20 or an erc1155. However, with underlying tokens, the vault might be 1155 while the underlying is erc20 - and this causes issues. When no vaultType is passed in it defaults to ERC20.

However that all being said, there is probably a way to handle this that is more clear with the business logic. I'll look into this

File: src/constants/misc.ts (hunk 2 above)
Comment (Rudi114) [reply src/constants/misc.ts:40]: We have some inconsistancies in the designs around this... "Setup Proxy" is capitalized for both words while "Deposit colateral" and "Deposit underlying" only have a single capitalized word.... 

I dont care which way we go but we should pick one for consistancy

File: pages/create-position/[collateralId]/open/index.tsx (hunk 3 above)
Comment (Rudi114) [reply pages/create-position/[collateralId]/open/index.tsx]: Good catch. This has been fixed

File: src/components/custom/create-position-underlying/index.tsx [hunk 8]
```diff
@@ -141,39 +147,38 @@ export const CreatePositionUnderlying: React.FC<CreatePositionUnderlyingProps> =
...
+      setLoading(true)
+      await buyCollateralAndModifyDebtERC1155({
+        vault: collateral.vault.address,
+        token: collateral.address ?? '',
+        tokenId: Number(collateral.tokenId),
+        deltaDebt: _fiatAmount,
+        virtualRate: collateral.vault.virtualRate,
+        fCashAmount: getHumanValue(fCashAmount, 53), // 41 puts us at 18decimals, 53 puts us at 6 decimals
+        minImpliedRate: minImpliedRate,
+        underlierAmount: _underlierAmount, //definitely correct */
```
Comment (whoislewys) [inline src/components/custom/create-position-underlying/index.tsx:175]: lol old comment?

File: src/hooks/useMinImpliedRate.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
+import { useWeb3Connected } from '@/src/providers/web3ConnectionProvider'
+import { contracts } from '@/src/constants/contracts'
+import { Contract } from 'ethers'
+
+export const useMinImpliedRate = (fCash: any, maxSlippage: any): any => {
+  const { appChainId, web3Provider } = useWeb3Connected()
+  const contract = new Contract(
+    contracts.USER_ACTIONS_FC.address[appChainId], //probably need to grab a different contract
```
Comment (whoislewys) [inline src/hooks/useMinImpliedRate.tsx:8]: did this end up being the contract?

File: src/hooks/useMinImpliedRate.tsx [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    const ratio = (proportion * RATE_PRECISION) / (RATE_PRECISION - proportion)
+    if (ratio < 0) {
+      console.log('Insufficient liquidity')
+    }
+    const logValue = Math.log(ratio / RATE_PRECISION)
+    return logValue * RATE_PRECISION
+  }
+
+  const interstToExchangeRate = (annualRate: any, timeToMaturity: any) => {
```
Comment (whoislewys) [inline src/hooks/useMinImpliedRate.tsx:29]: #nit `interestToExchangeRate`

**Review by whoislewys — COMMENTED** [review]
looks great!

haven't tested the underlying flow because i want to test underlying + newfeature flag together. 

take a look at these comments and hit me up when they're resolved and we can merge to develop to test the feat. flag

**Pull request merged**